Teacher resources on resource model

bethanyaconnor opened the pull request:
Step 3 of moving teacher resources onto the Resource model. Completes [PLAT-509](https://codedotorg.atlassian.net/browse/PLAT-509).

Steps for this will be:

1. (#39367) Create an association between resources and unit groups/scripts so that resources can be associated with a script or unit group.
2. (#39617 and #39605) Add seeding of ScriptsResource and UnitGroupsResource
3. (This PR) Replace the current way to add resources to a script with a new editor using the Resources model. Move all existing resources to the resources model on migrated scripts. Display resources from the resources model in the dropdown.

I'm sorry this PR ended up being so large but there's a lot of moving parts that really should happen at the same time: changing the UI, passing down the new data, and migrating the existing teacher resources. In addition, because there are a lot of files touched, a lot of tests needed to be updated, particularly in apps. The biggest changes are in React code for the script edit and course edit pages.

**Why only migrated scripts?**
For this dropdown, we'll be using the name of the resource, which isn't translateable (yet). For existing teacher resources, we mapped the type of the resource to a string that was translateable, but that system doesn't fit with the new models. If we did this work for in-use scripts, we would be losing those translations for that dropdown. [PLAT-865](https://codedotorg.atlassian.net/browse/PLAT-865) tracks the follow-up work here.

New UI on the script edit page:
https://user-images.githubusercontent.com/46464143/112667985-28850700-8e1b-11eb-92f0-ccfb5d575a7e.mov



## Testing story

unit tests, manual testing

## Deployment strategy

I'll need to double check that no more teacher resources are added before merging this PR. I don't believe that's a common thing so I'm hopeful if there are any updates, they will be small.

## PR Checklist:

<!--
  The final step! Before you create your PR, double-check that everything is in order.
  Change [ ] to [X] during creation to check boxes.
-->

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

davidsbailey commented on apps/src/code-studio/progress.js:
```diff
@@ -165,9 +165,12 @@ progress.renderCourseProgress = function(scriptData) {
+    : (scriptData.teacher_resources || []).map(([type, link]) => ({
+        type,
+        link
+      }));
```
storing different data types in the same field scares me a bit. could these be stored in separate fields? using the same field is (for me) harder to reason about, and it seems like using the wrong value of is_migrated could lead to a harder-to-debug type mismatch. I'm not 100% sure but I think this might have made it a bit easier to split up the PR, too.

Even if that sounds like a better solution to you, this is definitely a judgment call on whether it's worth doing... this code has an expiration date on it, so some added complexity could be ok, but on the other hand that date is still a couple months out. Happy to discuss :-)

bethanyaconnor replied:
hmm let me try that out. I think I was trying to avoid too much `migratedTeacherResources` kinds of things but with the two resource editors, that didn't really work out anyway

bethanyaconnor replied:
Took a stab at this in fcfcac9ad5a. Does this seem better? Part of me thinks it makes reasoning about the props simpler, like you said, but also the logic about which resources to use seems to leak through more. Thoughts?

davidsbailey replied:
thank you for trying this out! personally, having the logic "leak through" more seems to make the code a lot more readable to me. I would vote for this approach. Are there particular parts of this solution that are most troublesome to you?

bethanyaconnor replied:
Not particularly troublesome. If it's clearer, then I think it's a better approach. I'll update tests to match

davidsbailey commented on apps/src/lib/levelbuilder/course-editor/ResourcesEditor.js:
```diff
@@ -106,6 +106,7 @@ export default class ResourcesEditor extends Component {
           <div style={{marginBottom: 5}}>Preview:</div>
           <TeacherResourcesDropdown
             resources={resources.filter(x => !!x.type)}
+            useMigratedResources={false}
```
Why is this? it makes me think that the course editor will always use non-migrated resources, but earlier i see the course editor checking props.useMigratedResources.

bethanyaconnor replied:
This component is only used for old migrated resources -- it's the one that currently shows up on the course edit and unit edit pages. This whole file has an expiration date on it and won't ever use migrated resources.

Would it be worth it for me to move the "new" `ResourcesEditor`? Currently it lives in `lesson-editor/` but that increasingly seems like not the right place for it. I could move it as a follow-up to this PR

davidsbailey replied:
> This component is only used for old migrated resources

do you mean old, _un_-migrated resources? sorry to pick on a likely typo, but just want to be sure there's not a dimension to this that I'm missing.

> Would it be worth it for me to move the "new" ResourcesEditor? Currently it lives in lesson-editor/ but that increasingly seems like not the right place for it.

I think the hardest part for me in reading this code is figuring out what's old / unmigrated and what's new / migrated. Generally yes, some kind of distinction would be great, so that it's clear from the file name + path whether it is old or new.  If this file has an expiration date on it, then renaming the file with a prefix of Unmigrated, Legacy or Deprecated could be another option. I'm only suggesting that because I'm not quite sure what's on the table for what directory you'd move this to -- however moving it could be fine too if you have a vision for that.

davidsbailey replied:
looking at the rest of the PR again, I think I'd prefer the filename / component name to make the distinction, so that it's clear from wherever the component is instantiated that it is legacy / deprecated.

davidsbailey replied:
just one more thought to add here -- if we are consistent about naming the new components Migrated, then it seems ok not to rename the old components. :-)

bethanyaconnor replied:
> do you mean old, un-migrated resources? sorry to pick on a likely typo, but just want to be sure there's not a dimension to this that I'm missing.

Yes, sorry. And please pick on this -- I've been getting them backwards a bunch and it's really important to be clear here.

> I think I'd prefer the filename / component name to make the distinction

So in this case, rename this file/component to be `LegacyResourcesEditor`?

davidsbailey replied:
> So in this case, rename this file/component to be LegacyResourcesEditor?

That's what I was suggesting, however it looks like the name scheme you are generally using is that no prefix means legacy and "migrated" prefix means migrated, so if you are consistent about that them I'm ok with it.

davidsbailey commented on apps/src/lib/levelbuilder/script-editor/ScriptEditor.jsx:
```diff
@@ -71,7 +71,7 @@ class ScriptEditor extends React.Component {
     initialProjectWidgetVisible: PropTypes.bool,
     initialProjectWidgetTypes: PropTypes.arrayOf(PropTypes.string),
-    initialTeacherResources: PropTypes.arrayOf(resourceShape).isRequired,
+    initialTeacherResources: PropTypes.arrayOf(PropTypes.object).isRequired,
```
If you are going to keep supporting both data types on the same field, I would suggest defining both shapes, and using PropTypes.oneOf to combine them.

davidsbailey reviewed (commented):
I've got some initial comments and questions... will need to look again with a fresh brain tomorrow.

davidsbailey commented on apps/src/lib/levelbuilder/course-editor/CourseEditor.js:
```diff
@@ -263,17 +274,40 @@ export default class CourseEditor extends Component {
+              value={this.props.migratedResources.map(r => r.id)}
+            />
+          )}
+          {this.props.useMigratedResources ? (
```
If you did want to reduce the number of times you check useMigratedResources throughout the code, one idea would be that once you have set resources and migratedResources, to just check the contents of those props when deciding what to display, and trust that whatever code higher up the stack did not accidentally set both

dmcavoy commented:
> Why only migrated scripts?
For this dropdown, we'll be using the name of the resource, which isn't translateable (yet). For existing teacher resources, we mapped the type of the resource to a string that was translateable, but that system doesn't fit with the new models. If we did this work for in-use scripts, we would be losing those translations for that dropdown. PLAT-865 tracks the follow-up work here.

Is the teacher resources dropdown used in CSF? Otherwise it seems like the translation here might not be a huge loss since so much of CSD/CSP are not translated anyway

dmcavoy commented on apps/src/code-studio/components/progress/ScriptOverview.jsx:
```diff
@@ -173,11 +177,13 @@ class ScriptOverview extends React.Component {
               viewAs={viewAs}
               isRtl={isRtl}
               resources={teacherResources}
+              migratedResources={migratedTeacherResources}
```
nit: would it be worth being consistent with the name? either migratedTeacherResources or migratedResources. My vote is for the first

bethanyaconnor replied:
That makes sense to me. I'll update

dmcavoy commented on apps/src/lib/levelbuilder/script-editor/ScriptEditor.jsx:
```diff
@@ -817,17 +825,35 @@ class ScriptEditor extends React.Component {
+                />
+              </div>
+            )}
+            {this.props.isMigrated && (
```
It looks like maybe the code for deciding which editor to show and the parts of the editor might be the same between Course and Script edit? Is it worth creating a component to share?

bethanyaconnor replied:
oh good thought -- I could actually roll this logic into `ResourcesEditor` and have that handle figuring out what to show. That would have the added benefit of not having to preview the dropdown here too

bethanyaconnor replied:
Done in 1caab4299b9. I think it consolidates some of this logic but let me know if you disagree.

The only downside is that the editor looks a bit cluttered on migrated scripts but we can clean this up later
<img width="1096" alt="Screen Shot 2021-03-31 at 10 35 21 AM" src="https://user-images.githubusercontent.com/46464143/113186804-f4d82180-920c-11eb-9a66-3713f1e66427.png">

dmcavoy commented:
Pulled it locally and tried it out. Looks great!

dmcavoy commented:
It looks like you can add a resource that is marked as student facing to the teacher resources dropdown. How do we feel about that?

bethanyaconnor commented:
> It looks like you can add a resource that is marked as student facing to the teacher resources dropdown. How do we feel about that?

This seems like something we wouldn't necessarily want to prohibit but we could make it more obvious? There are a fair number of resources that have a teacher and student version and it would be pretty easy to misclick

dmcavoy commented:
> > It looks like you can add a resource that is marked as student facing to the teacher resources dropdown. How do we feel about that?
> 
> This seems like something we wouldn't necessarily want to prohibit but we could make it more obvious? There are a fair number of resources that have a teacher and student version and it would be pretty easy to misclick

Ya I didn't know with use adding a student resource dropdown if we would decide which items to show it which dropdown would be based on the audience they are assigned to. I think there is value it allowing for you to add whatever resource you want so probably need to two versions of the editor - one for teachers and one for students

bethanyaconnor commented:
> > > It looks like you can add a resource that is marked as student facing to the teacher resources dropdown. How do we feel about that?

> > This seems like something we wouldn't necessarily want to prohibit but we could make it more obvious? There are a fair number of resources that have a teacher and student version and it would be pretty easy to misclick

> Ya I didn't know with use adding a student resource dropdown if we would decide which items to show it which dropdown would be based on the audience they are assigned to. I think there is value it allowing for you to add whatever resource you want so probably need to two versions of the editor - one for teachers and one for students

Ah makes sense! I like having two editors but I was thinking of adding the audience to the search dropdown so it's easier to know which one is being selected. But I'll do that in a different PR

davidsbailey commented on apps/src/code-studio/components/progress/ScriptOverview.jsx:
```diff
@@ -37,7 +37,8 @@ class ScriptOverview extends React.Component {
     excludeCsfColumnInLegend: PropTypes.bool.isRequired,
-    teacherResources: PropTypes.arrayOf(resourceShape).isRequired,
+    teacherResources: PropTypes.arrayOf(resourceShape),
+    migratedTeacherResources: PropTypes.arrayOf(PropTypes.object),
```
sorry if this is premature, but can you please add a more specific shape for this PropType instead of object?

bethanyaconnor replied:
Yep -- just pushed in 1e2176dd686

bethanyaconnor commented:
@davidsbailey @dmcavoy I _think_ I addressed all the feedback, though I definitely could've missed something. Mind taking another look when you get a chance?

dmcavoy commented on apps/src/lib/levelbuilder/course-editor/CourseEditor.js:
```diff
@@ -263,17 +274,30 @@ export default class CourseEditor extends Component {
-            Select the Teacher Resources buttons you'd like to have show up on
-            the top of the course overview page
+            <div>
+              Select the Teacher Resources buttons you'd like to have show up on
```
I know this isn't new but this text looks like its out of date. Looks like when we updated the course teacher resource to be a dropdown like script we didn't update this text to match.

bethanyaconnor replied:
It's the same text as is on the script edit page. I'm happy to change it but I'm not sure what the best wording is?

dmcavoy replied:
I think "Select the Teacher Resources you'd like to have show up in the dropdown at the top of the course overview page"

davidsbailey commented on apps/src/lib/levelbuilder/shapes.jsx:
```diff
@@ -74,6 +74,7 @@ export const activityShape = PropTypes.shape({
 });
 
 export const resourceShape = PropTypes.shape({
+  id: PropTypes.number,
```
because this does not have "migrated" in the name, I think it would be helpful to add a comment indicating that this represents a migrated resource. it might be even better to rename it to migratedResourcesShape, but  that can be optional :-)

bethanyaconnor replied:
This is used in other places so I'd prefer to not have to change them all. I'll add a comment though

davidsbailey commented on apps/src/sites/studio/pages/courses/edit.js:
```diff
@@ -1,8 +1,12 @@
 import CourseEditor from '@cdo/apps/lib/levelbuilder/course-editor/CourseEditor';
+import resourcesEditor, {
+  initResources
+} from '@cdo/apps/lib/levelbuilder/lesson-editor/resourcesEditorRedux';
```
Does all of resourcesEditorRedux deal only with migrated resources? If so, it could be helpful to note that in a comment in that file.

davidsbailey commented on dashboard/config/scripts_json/coursea-2021.script_json:
```diff
@@ -14,23 +14,13 @@
     "new_name": null,
     "family_name": "coursea",
-    "serialized_at": "2021-03-24 21:38:25 UTC",
+    "serialized_at": "2021-03-30 16:02:40 UTC",
```
I'd recommend removing these `serialized_at` lines from your diffs, since they have a high chance of merge conflict, but it's up to you to do whatever lets you get this merged in the easiest

bethanyaconnor replied:
Yeah, that's definitely part of the merge conflicts that we're seeing. I'll remove them when I regenerate the diffs

davidsbailey reviewed (approved):
Looks good to me! The PR is still quite large, but I am now having a _much_ easier time of answering "what is it?" when I look at an isolated part of this PR and try to figure out what is legacy and what is migrated. Thanks for your patience! My remaining requests are just to add comments.

dmcavoy commented on dashboard/app/models/resource.rb:
```diff
@@ -97,9 +97,20 @@ def summarize_for_lesson_edit
     }
   end
 
+  def summarize_for_script_unit_group_overview
```
nit: The name of this method is confusing to me. Does this mean it is summarizing for both the script overview and unit group overview?

davidsbailey replied:
I had the same confusion. I think `summarize_for_script_or_unit_group_overview` or just `summarize_for_overview` would make more sense to me

dmcavoy replied:
or `summarize_for_teacher_resources_dropdown`

davidsbailey replied:
> or `summarize_for_teacher_resources_dropdown`

that sounds better!

bethanyaconnor replied:
Yes, I like it. Thanks!

dmcavoy reviewed (approved):
Awesome work on this Bethany! This was a lot to get together and working but I think it will be a big win.
